proposals/0529-th-quotes-as-patterns.rst

```diff
@@ -0,0 +1,356 @@
+
+The solution is simple: let's allow quotes too!
+This would allow:
+
+.. code-block:: diff
+
+         Just (Refl :: d :~: Exp)
+  -        | AppE (VarE m) eInner <- d
+  +        | [e| $(VarE m) $eInner |] <- d
+           , m == 'unqMarker
+           -> do ...
+
+or even going further:
+
+.. code-block:: diff
+
+         Just (Refl :: d :~: Exp)
+  -        | AppE (VarE m) eInner <- d
+  -        , m == 'unqMarker
+  +        | [e| unqMarker $eInner |] <- d
+           -> do ...
+
+In this way, we also avoid the use of the AST.
+
+Note we do have quotes *of patterns* today (``[p| ... |]``), but that is orthogonal.
+This is quotes *as patterns*, the type of syntax being quoted doesn't matter and could be anything.
+The point is the quotes are in negative position.
+
+With this change put together, the hope is that a significant portion of TH out in the wild is going to be more stable across GHC versions.
+This is all accomplished without trying to minimize TH AST changes, which is quite a hopeless task and also a perverse incentive for the rest of language development.
+
+Proposed Change Specification
+-----------------------------
+
+Quotes as patterns
+~~~~~~~~~~~~~~~~~~
+
+With the new extension ``TemplateHaskellQuotesAsPatterns``, slightly modified quotes are usable in pattern position.
+
+The first difference is that quotes as pattern match raw syntax, not (monadic) actions producing syntax.
+The tying rules are as follows:
+
+- ``[e| ... |]`` matches ``Exp``
+- ``[p| ... |]`` matches ``Pat``
+- ``[t| ... |]`` matches ``Type``
+- ``[d| ... |]`` matches ``Dec``
+
+The second differences is that splices within these quotes contain patterns instead of expressions::
+
+  p is in <pat>
+  --------------------------------
+  [| ... $(p) ... |] is in <apat>
+
+The third and final difference is that names in quotes must all be uses, never bindings.
+
+Optional: Fine-grained Quotation constraints
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. _`Proposal 246`: https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0246-overloaded-bracket.rst
+
+*This is optional but fits well with the rest.*
+
+`Proposal 246`_ made it so that quotes are polymorphic, e.g. ``[| 1 + 1 |] :: Quote m => m Exp``.
+The ``Quote`` class has a ``newName`` method, and is just used when binding local variables.
+Relax the rules so that TH Quotes only impose a ``Quote`` constraint when ``newName`` is in fact needed.
+
+
+Examples
+--------
+
+#. This is allowed::
+
+     f [| $(x) $(y) |] = ...
+
+     =>
+
+     f (AppE x y) = ...
+
+#. This is conditionally allowed::
+
+     f [| name |] = ...
+
+     =>
+
+     f (VarE __n) | __n == 'name = ...
+
+   on ``'name`` being bound in the scope the bracket is written in (bound lexically, not dynamically at the splice site).
+
+#. This is not allowed::
+
+     f [| \x -> x |] = ...
+
+   It is disallowed because the first ``x`` in the quote is a binding not a use.
+
+Effect and Interactions
```

Comment:
I'm assuming that the `TemplateHaskell` will not imply the new extension, though it does imply `TemplateHaskellQuotes`?

Reply:
Yes. Added.